### DISCUSSION TITLE
Question: insertBefore() loses element focus?

### DISCUSSION THREAD

**Discussion by jenstroeger** [jquery/jquery#4923] [Q&A]
It looks like the “Sign up” link to the [jQuery Forum](https://forum.jquery.com/using-jquery) is broken and Freenode requires a login/auth for IRC. So, here I am with a question…

Is it possible that [insertBefore()](https://api.jquery.com/insertBefore/) (and similar functions) when moving an element within the DOM make the moved element (or an element in the subtree) lose its focus?

**Comment by jenstroeger** [comment]
**Answer**: It is possible.

Suppose `#elem` (or one of its children) has the [focus](https://developer.mozilla.org/en-US/docs/Web/API/Element/focus_event). Then
```javascript
$("#elem").insertBefore("#other")
```
blurs on `#elem` so that it can be moved around the DOM. In contrast
```javascript
$("#other").insertAfter("#elem")
```
keeps the focus on `#elem`. I suspect that similar behavior could be observed with [before()](https://api.jquery.com/before/) and [after()](https://api.jquery.com/after/) as well.

Might be worth mentioning in the docs?

Ok to close this question.

**Reply by dmethvin** [reply]
The thing getting moved is usually taken out of the DOM temporarily in order to be moved in the DOM tree. That applies to a lot of jQuery methods, not just `insertBefore`. Also note that the behavior of `$("#elem").insertBefore(".severalOtherElements")` can't possibly keep focus on all the copies. So most likely the problem is that the page is trying to move elements the user has focused, which can't be fixed inside jQuery. You can see the same problem with bare DOM operations: https://jsbin.com/mujodujegu/edit?html,console,output

It is not a good idea to programmatically mess with the user's focus, and by extension any operation that would modify DOM that has focus (moving, wrapping, removing, shifting visually on the page) is probably not a good UX. You can try to patch up focus by checking whether `document.activeElement` is contained by whatever DOM is moving and set it back after the move, but I don't think that addresses the UX issue.

**Reply by jenstroeger** [reply]
For context: I asked the question because elements had `focus` and `blur` event listeners registered, and by moving a parent element within the DOM (i.e. taking it out and inserting it someplace else) these events would fire. So, I was looking for a way to avoid the events from triggering — either by deregistering the listeners or by trying to keep the focus on the element.

**Reply by mgol** [reply]
Since the element loses focus, it looks like the fact these event handlers fire is correct?

**Reply by Mp4official** [reply]
Yes

**Comment by mgol** [comment]
Converted to a discussion. If you have documentation suggestions, it's best to submit a PR at https://github.com/jquery/api.jquery.com.
